keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,229 @@
+  completed.
+* Even though we cannot reconstruct this field immediately, this field is
+  initialized to empty, meaning "unknown", which truthfully reflects the state.
+* Putting it in the spec is even more unnatural. Spec describes the desired
+  state the system should eventually reach in the future, while this field is
+  about states in the history.
+* Keeping it in the status is consistent with the [CRD.status.StoredVersions][]
+  API.
+
+[CRD.status.StoredVersions]:https://github.com/kubernetes/kubernetes/blob/697c2316faaabae8ef8371032b60be65d7795e68/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L305
+  
+## Storage migration triggering controller
+
+We will add a controller that monitors the discovery documents and the
+storageStates to trigger storage version migration.
+
+The controller fetches the discovery document periodically. For each resource,
+if the storageState doesn't exist for the resource, the controller
+  * stops any ongoing migration of this resource by deleting existing
+    [storageVersionMigrations][],
+  * creates a new storageVersionMigration for this resource,
+  * creates the storageState for the resource, setting
+    .status.storedVersionsHash to ["Unknown"], and status.StorageVersionHash to
+    the value shown in the discovery document.
+
+If the storageState exists, the controller compares the storageVersion in the
+discovery document with that recorded in the storageState. If they are the same,
+the controller simply updates the status.lastHeartbeatTime and is done.
+Otherwise, the controller
+  * stops any ongoing migration of this resource by deleting existing
+    [storageVersionMigrations][],
+  * creates a new storageVersionMigration for this resource,
+  * as a single operation, updates the status.lastHeartbeatTime, sets
+    storageState.status.StorageVersionHash to the value shown in the discovery
+    document, and appends the hash to the storageState.status.storedVersionHashes.
+
+The refresh rate of the discovery document needs to satisfy two conflicting
+goals: *i)* not causing too much traffic to apiservers, and *ii)* not missing
+storage version configuration changes. More precisely, the storage version of a
+certain resource should not change twice within a refresh period. 10 minutes is
+a good middle ground.
+
+The controller uses an informer to monitor all [storageVersionMigrations][]. If a
+migration completes, the controller finds the storageState of that resource,
+updates its status.storedVersionHashes to equal its status.storageVersionHash.
+
+If the controller crash loops and misses storage version changes, the
+storageState.status.storedVersionHashes can be stale. Thus, when the controller
+bootstraps, for every storageState object, it checks the
+.status.lastHeartbeatTime, if the timestamp is more than 10 minutes old, it
+recreates storageState object with storedVersionHashes set to ["Unknown"].
```

Comment:
This seems less than ideal.
* It's easy to imagine a failed upgrade+rollback that takes less than 10 minutes, yet leaves behind a few objects with a different storage version.
* It's also easy to imagine an occasional failure to update within the 10 minute window.

I think in my original design I was tracking this directly in the discovery document :/

Reply:
> * It's easy to imagine a failed upgrade+rollback that takes less than 10 minutes

Cluster admin needs to manually reset the state in this case. It's noted in the "implications to cluster admin" section.

> * It's also easy to imagine an occasional failure to update within the 10 minute window

The migration triggering controller will reset the state. Correctness is guaranteed, but the migration will be done over and over.

> I was tracking this directly in the discovery document :/

That would require the apiserver to persist the discovery document, right?

Can you put up with these caveats?

Reply:
I'm not actually advocating for modifying the discovery document.

Until we make apiserver coordinate schema changes, there's literally no way to get a push notification (i.e., watch) on schema changes, so there's not much to be done.

So, although this is not great, I don't know if one can realistically do better.